pandas/core/groupby.py

```diff
@@ -1923,8 +1923,16 @@ def __init__(self, index, grouper=None, obj=None, name=None, level=None,
 
             # a passed Categorical
             elif isinstance(self.grouper, Categorical):
+
+                # must have an ordered categorical
+                if self.sort:
+                    if not self.grouper.ordered:
+                        raise ValueError("cannot sort by an unordered Categorical in the grouper\n"
```

Comment:
Ok, the much more sensible workaround would be `df["B"]= df["A"].astype('category', ordered=True)` and it even does the right thing if "option 4" happens (if not, this would result in differently ordered groups: A by lexi-sort and B by ordee of appearance...)